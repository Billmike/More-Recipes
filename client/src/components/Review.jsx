import React from 'react';
import moment from 'moment';

const Review = (props) => {
  return (
    <div className="row custom-div">
      <div className="col-md-1">
        <img className="review-image" />
      </div>
      <div className="col-md-11 review-box">
        <p className="reviewer-name inline-p">{props.user}</p>
        <span
          className="review-time">
          {moment(props.review.createdAt, 'YYYY-MM-DD HH:mm Z').fromNow()}
        </span>
        <hr />
        <p className="review-content">{props.review.content}</p>
      </div>
    </div>
  );
};

export default Review;
