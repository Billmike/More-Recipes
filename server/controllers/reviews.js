import db from '../models/index';
import sendMail from './sendMail';
import errorMessage from '../errorHandler/errorMessage';

const { User } = db;
const reviews = db.Review;
const recipes = db.Recipe;

/**
  * Represents the Review class
  *
  * @class
  */

class Review {
/**
  * Represents the method that adds a review
  * @method
  *
  * @param { object } req - The request object
  * @param { object } res - The response object
  *
  * @returns { object } Returns the review object
  *
  */

  static addReviews(req, res) {
    return recipes.findById(req.params.recipeId)
      .then((foundRecipe) => {
        if (!foundRecipe) {
          return res
            .status(404)
            .json({ message: 'Recipe not found.' });
        }
        if (foundRecipe.owner === req.userId) {
          return res
            .status(403)
            .json({
              message: 'You cannot review your own recipe.'
            });
        }
        User.findById(foundRecipe.owner)
          .then((user) => {
            reviews.create({
              userId: req.userId,
              recipeId: req.params.recipeId,
              content: req.body.content,
            }).then((review) => {
              const mailOptions = {
                from: process.env.EMAIL_ADDRESS,
                to: user.email,
                subject: `Hi ${user.username}. Your recipe has been reviewed`
              };
              sendMail.sendMail(mailOptions, (err) => {
                if (err) {
                  return res.json({
                    message: err.message
                  });
                }
              });
              return res.status(201).json({
                message: 'Review successfully posted',
                reviewData: {
                  content: review.content
                }
              });
            });
          });
      }).catch(() => {
        return res
          .status(500)
          .json({
            message: errorMessage
          });
      });
  }
}

export default Review;
