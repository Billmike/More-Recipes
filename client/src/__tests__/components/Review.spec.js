import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Review from '../../components/Review';
import reviews from '../fixtures/reviews';

test('Should render review component correctly', () => {
  const wrapper = shallow(<Review review={reviews} />);
  expect(wrapper.find('div').length).toBe(2);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
