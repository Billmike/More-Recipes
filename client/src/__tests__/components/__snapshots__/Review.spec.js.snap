// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render review component correctly 1`] = `
<div
  className="row custom-div"
>
  <div
    className="col-md-11 review-box"
  >
    <p
      className="reviewer-name inline-p"
    />
    <span
      className="review-time"
    >
      Invalid date
    </span>
    <hr />
    <p
      className="review-content"
    >
      Amazing recipe you got fam!
    </p>
  </div>
</div>
`;
